### PR TITLE
Range detection and base classes

### PR DESCRIPTION
**Pull request by rhaDHI**

### CODE REVIEW

File: anomalydetection/detectors.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import pandas as pd
+
+
+class BaseDetector:
+    def __init__(self):
+        pass
+
+    def fit(self, data: pd.Series):
```
Comment (laurafroelich) [inline anomalydetection/detectors.py:8]: Should we consider following sklearn, having both x and y as input, with y being optional? or is the plan to limit this package to supporting only unsupervised anomaly detection methods? I guess that would also be fine, for supervised approaches there are already many tools.

File: anomalydetection/detectors.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    def fit(self, data):
+        for detector in self._detectors:
+            detector.fit(data)
+        return self
+
+    def detect(self, potentially_abnormal_data):
+        detected_anomalies = [False] * len(potentially_abnormal_data)
+        for detector in self._detectors:
+            detected_anomalies |= detector.detect(potentially_abnormal_data)
```
Comment (laurafroelich) [inline anomalydetection/detectors.py:37]: Perhaps a voting strategy would work better? With more detectors, risk of false positive detections will increase

File: anomalydetection/detectors.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            return data < self._min
+
+        if self._min is None:
+            return data > self._max
+
+        return (data < self._min) | (data > self._max)
+
+    def _validate_fit(self):
+        if self._min is None and self._max is None:
+            raise ValueError("Please call fit() before detect() or "
```
Comment (laurafroelich) [inline anomalydetection/detectors.py:88]: Could we use a custom exception instead?

File: tests/range_detection.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    normal_data = np.array([0, np.nan, 1, 0, 2, np.nan, 3.14, 4])
+    abnormal_data = np.array([-1, np.nan, 2, np.nan, 1, 0, 4, 10])
+    expected_anomalies = np.array([True, False, False, False, False, False, True, True])
+    assert len(expected_anomalies) == len(abnormal_data)
+    return normal_data, abnormal_data, expected_anomalies
+
+
+@pytest.fixture
+def range_data_series(range_data):
+    normal_data, abnormal_data, expected_anomalies = range_data
```
Comment (laurafroelich) [inline tests/range_detection.py:19]: range_data needs paranthesis for function to be called

File: tests/range_detection.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def test_range_detector(range_data_series):
+    data, _, _ = range_data_series
+
+    detector = RangeDetector(0, 2)
+    anomalies = detector.detect(data)
+
+    assert len(anomalies) == len(data)
+    assert sum(anomalies) == 2
```
Comment (laurafroelich) [inline tests/range_detection.py:31]: I would prefer a test that tests detected anomalies are exactly equal to expected ones

File: tests/range_detection.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def test_range_detector_autoset(range_data_series):
```
Comment (laurafroelich) [inline tests/range_detection.py:34]: Similar comments as above:
Paranthesis needed to call range_data_series

I would prefer exact tests that assert that detected anomalies are equal to those expected

File: tests/range_detection.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    anomalies = RangeDetector(min_value=3).detect(data)
+    assert sum(anomalies) == 4
+
+    anomalies = RangeDetector(max_value=3).detect(data)
+    assert sum(anomalies) == 2
+
+
+def test_range_detector_pipeline(range_data_series):
+    normal_data, abnormal_data, expected_anomalies = range_data_series
```
Comment (laurafroelich) [inline tests/range_detection.py:45]: range_data_series needs paranthesis to be called

File: anomalydetection/detectors.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        pass
+
+    def fit(self, data: pd.Series):
+        """ Set detector parameters based on data. """
+        self.validate(data)
+        return self
+
+    def detect(self, data: pd.Series):
+        """ Detect anomalies in data. """
+        return pd.Series([False] * len(data), index=data.index)
```
[the remark this answers is not in the document]
Comment (rhaDHI) [reply anomalydetection/detectors.py:15]: Would you throw a NotImplementedError exception here then or just pass?

File: anomalydetection/detectors.py (hunk 1 above)
Comment (rhaDHI) [reply anomalydetection/detectors.py:8]: I am not sure yet. detect should probably also be predict then. Let's decide later if we need to mix with sklearn

File: anomalydetection/detectors.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (laurafroelich) [reply anomalydetection/detectors.py:15]: I would raise a NotImplementedError

File: anomalydetection/detectors.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhaDHI) [reply anomalydetection/detectors.py:8]: I don't understand why

File: anomalydetection/detectors.py (hunk 2 above)
Comment (rhaDHI) [reply anomalydetection/detectors.py:37]: So you are proposing &= instead ? Or changing from bool to int to count the votes?

File: tests/range_detection.py (hunk 4 above)
Comment (rhaDHI) [reply tests/range_detection.py:19]: If we do then we get: TypeError: 'tuple' object is not callable

File: anomalydetection/detectors.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+    def validate(self, data):
+        if not isinstance(data, pd.Series):
+            raise ValueError("Input data must be a pandas.Series.")
```
[the remark this answers is not in the document]
Comment (rhaDHI) [reply anomalydetection/detectors.py]: Yep 👍

File: anomalydetection/detectors.py (hunk 2 above)
Comment (rhaDHI) [reply anomalydetection/detectors.py:38]: I moved to issue #3

**Pull request merged**